src/effects/native/gatereffect.cpp

```diff
@@ -0,0 +1,236 @@
+    triplet->setMinimum(0);
+    triplet->setMaximum(1);
+
+    EffectManifestParameter *invert = manifest.addParameter();
+    invert->setId("invert");
+    invert->setName("Invert");
+    invert->setDescription("");
+    invert->setControlHint(
+        EffectManifestParameter::ControlHint::TOGGLE_STEPPING);
+    invert->setSemanticHint(EffectManifestParameter::SemanticHint::UNKNOWN);
+    invert->setUnitsHint(EffectManifestParameter::UnitsHint::UNKNOWN);
+    invert->setDefault(0);
+    invert->setMinimum(0);
+    invert->setMaximum(1);
+
+    return manifest;
+}
+
+GaterEffect::GaterEffect(EngineEffect* pEffect,
+                             const EffectManifest& manifest)
+        : m_pRateParameter(pEffect->getParameterById("rate")),
+          m_pShapeParameter(pEffect->getParameterById("shape")),
+          m_pQuantizeParameter(pEffect->getParameterById("quantize")),
+          m_pTripletParameter(pEffect->getParameterById("triplet")),
+          m_pInvertParameter(pEffect->getParameterById("invert")) 
+{
+    Q_UNUSED(manifest);
+}
+
+GaterEffect::~GaterEffect() {
+}
+
+void GaterEffect::processChannel(const ChannelHandle& handle,
+                                GaterGroupState* pState,
+                                const CSAMPLE* pInput, CSAMPLE* pOutput,
+                                const unsigned int numSamples,
+                                const unsigned int sampleRate,
+                                const EffectProcessor::EnableState enableState,
+                                const GroupFeatureState& groupFeatures) { 
+                                    
+    Q_UNUSED(handle);
+                                    
+    const double rate  = m_pRateParameter->value();
+    const double shape  = m_pShapeParameter->value();
+
+    // Used to divide a beat by a power of 2
+    int divider = pow(2, floor(3*rate));
+
+    // Attack & Release increase with the divider factor
+    double attackInc = kBaseAttackInc*divider;
+    double releaseInc = kBaseReleaseInc*divider;
+
+    // Get channel specific state variable
+    unsigned int timePosition    = pState->timePosition;
+    unsigned int holdCounter     = pState->holdCounter;
+    double       gain            = pState->gain;
+    GaterGroupState::State state = pState->state;
+
+    // Use to keep track of the beginning of beats when playing with the rate
+    unsigned int beatPeriod = 1;
+    if (enableState == EffectProcessor::ENABLING) {        
+        // Start with the gate closed
+        state = GaterGroupState::IDLE;
+        gain = 0;
+
+        // Set initial time from current beat position, or 0 if no beat grid
+        if (groupFeatures.has_beat_length_sec && groupFeatures.has_beat_fraction) {
+            beatPeriod   = groupFeatures.beat_length_sec*sampleRate;
+            timePosition = groupFeatures.beat_fraction*beatPeriod;
+        } else {
+            timePosition = 0;   
+            beatPeriod   = sampleRate; // 1 second
```

Comment:
Correct, this is 1 second in _sample frames_, but not when counting individual samples like the engine does. The engine design is somehow broken by indexing total number of samples from (implicitly) 2 channels instead of sample frames with interleaved samples. Indexing individual samples is dependent on the number of channels, actually 2. This design prevents processing of multiple channels or optimizations for mono signals.

It is not clear to me where and for which of the parameters you implicitly take this magic number of 2 channels into account. Maybe it would help to explicitly define a constexpr kNumberOfChannels = 2 and use it where applicable. In the next step you can rewrite your code to minimize usage of this constant. It should only be needed initially to transform the input parameters (divide by kNumberOfChannels) and finally when you actually need a "sample index" (multiply by kNumberOfChannels, adding 1 for the second channel).

If you try to keep your calculations independent of the number of channels and already use "sample frames" for the internal calculations your code becomes more obvious and migration will be easier, once the engine interfaces have been fixed.

Not your fault, this is a legacy issue ;)

Reply:
Consequently all calculations should be dependent only on the sample rate, but not on the number of channels. Multiplication/division by 2 is fast anyway, so I would recommend to strive for clear code instead of maximizing the performance for a broken interface design.